src/ops/constant.rs

```diff
@@ -215,9 +215,8 @@ pub trait CustomConst:
     /// An identifier for the constant.
     fn name(&self) -> SmolStr;
 
-    /// Returns the type of the constant.
-    // TODO it would be good to ensure that this is a *classic* CustomType not a linear one!
-    fn custom_type(&self) -> CustomType;
+    /// Check the value is a valid instance of the provided type.
+    fn check_type(&self, typ: &CustomType) -> Result<(), ConstTypeError>;
```

Comment:
part one: yes

part two: yes, but with some wrapping i think, will do

Reply:
ah no actually, CustomCheckFail is only there as a catch all in case the existing error modes don't match up, in most cases `ValueCheckFail` is the correct mode and can be reused.

Reply:
If there's no way for a CustomConst to synthesize its type, there's no way to tell that a CustomConst is Hashable or otherwise. If we don't want to add a synthesizing method, do we want to add a method returning a `TypeTag` ?

Reply:
We don't *need* such (#322 will work without), but that doesn't mean we shouldn't have it ;). I guess it could return Option<TypeTag>, and maybe should, whereupon we can't rely on it, so it doesn't reduce complexity (we still have to handle all the cases).